lib/utils/platform/iot_threads_afr.c

```diff
@@ -0,0 +1,357 @@
+
+bool IotMutex_Create( IotMutex_t * pNewMutex , bool recursive )
+{
+    IotLogDebug( "Creating new mutex %p.", pNewMutex );
+    struct iot_mutex_internal * pMutex = ( struct iot_mutex_internal * ) pNewMutex;
+    
+    if ( recursive  )
+    {
+        ( void ) xSemaphoreCreateRecursiveMutexStatic( & pMutex->xMutex );
+    }
+    else 
+    {
+        ( void ) xSemaphoreCreateMutexStatic( & pMutex->xMutex );
+    }
+    
+    pMutex->type = IOT_THREAD_MUTEX_DEFAULT; /* remember the type of mutex */
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Destroy( IotMutex_t * const pMutex )
+{
+    struct iot_mutex_internal * internalMutex = ( struct iot_mutex_internal * ) pMutex;
+
+    vSemaphoreDelete( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+}
+
+/*-----------------------------------------------------------*/
+
+bool prIotMutexTimedLock(IotMutex_t * const pMutex, TickType_t timeout)
+{
+    struct iot_mutex_internal * internalMutex = ( struct iot_mutex_internal * ) pMutex;
+    BaseType_t  lockResult;
+
+    IotLogDebug( "Locking mutex %p.", internalMutex );
+
+    /* Call the correct FreeRTOS mutex take function based on mutex type. */
+    if( internalMutex->type == IOT_THREAD_MUTEX_NORMAL )
+    {
+        lockResult = xSemaphoreTake( ( SemaphoreHandle_t ) & internalMutex->xMutex, timeout);
+    }
+    else
+    {
+        lockResult = xSemaphoreTakeRecursive( ( SemaphoreHandle_t ) & internalMutex->xMutex, timeout );
+    }
+  
+    return ( lockResult == pdTRUE );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Lock( IotMutex_t * const pMutex )
+{
+    prIotMutexTimedLock( pMutex, portMAX_DELAY );
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotMutex_TryLock( IotMutex_t * const pMutex )
+{
+    return prIotMutexTimedLock( pMutex, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Unlock( IotMutex_t * const pMutex )
+{
+    struct iot_mutex_internal * internalMutex = ( struct iot_mutex_internal * ) pMutex;
+
+    IotLogDebug( "Unlocking mutex %p.", internalMutex );
+    
+    /* Call the correct FreeRTOS mutex unlock function based on mutex type. */
+    if( internalMutex->type == IOT_THREAD_MUTEX_NORMAL )
+    {
+        ( void ) xSemaphoreGive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+    else
+    {
+        ( void ) xSemaphoreGiveRecursive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_Create( IotSemaphore_t * const pNewSemaphore,
+                             uint32_t initialValue,
+                             uint32_t maxValue )
+{
+    struct iot_sem_internal * internalSemaphore = ( struct iot_sem_internal * ) pNewSemaphore;
+    IotLogDebug( "Creating new semaphore %p.", pNewSemaphore );
+
+    if( maxValue > ( uint32_t ) SEM_VALUE_MAX )
```

Comment:
`SEM_VALUE_MAX` is a POSIX constant.  Probably don't have to check if `maxValue` is too big on FreeRTOS (can be any `UBaseType_t`.

Reply:
Just removed the check, I think any reference to POSIX should be avoided of course.